src/main/java/com/uid2/admin/job/JobDispatcher.java

```diff
@@ -2,39 +2,55 @@
 
 import com.uid2.admin.job.model.Job;
 import com.uid2.admin.job.model.JobInfo;
+import com.uid2.admin.store.Clock;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.Metrics;
 import io.vertx.core.logging.Logger;
 import io.vertx.core.logging.LoggerFactory;
 
+import java.time.temporal.ChronoUnit;
 import java.util.*;
 import java.util.concurrent.*;
 import java.util.stream.Collectors;
 
 public class JobDispatcher {
-    private static class Loader {
-        public static final JobDispatcher INSTANCE = new JobDispatcher();
-    }
-
     private static final Logger LOGGER = LoggerFactory.getLogger(JobDispatcher.class);
 
-    private final ExecutorService JOB_EXECUTOR = Executors.newSingleThreadExecutor();
-    private final Queue<Job> JOB_QUEUE = new ConcurrentLinkedQueue<>();
-    private final Object JOB_LOCK = new Object();
+    private final ExecutorService jobExecutor = Executors.newSingleThreadExecutor();
+    private final Queue<Job> jobQueue = new ConcurrentLinkedQueue<>();
+    private final Object jobLock = new Object();
+
+    private final String id;
+    private final int intervalMs;
+    private final int maxRetries;
+    private final Clock clock;
+
     private boolean started = false;
+    private Job currentJob = null;
     private ScheduledExecutorService scheduler;
-    private Job currentJob;
 
-    private JobDispatcher() {}
-
-    public static JobDispatcher getInstance() {
-        return Loader.INSTANCE;
+    public JobDispatcher(
+            String id,
+            int intervalMs,
+            int maxRetries,
+            Clock clock) {
+        this.id = id;
+        this.intervalMs = intervalMs;
+        this.maxRetries = maxRetries;
+        this.clock = clock;
+
+        Gauge.builder("uid2.job_dispatcher.execution_duration_ms", this::getExecutionDuration)
+                .tag("job_dispatcher", id)
+                .description("gauge for " + id + " execution time")
+                .register(Metrics.globalRegistry);
```

Comment:
so prometheus will poll for this execution_duration_ms value periodically via the Gauge mechanism? would alert be based on such gauge? (e.g. if prometheus always poll at the time when currnetJob is null then this Gauge wouldn't be helpful? ) 

or should we always store the mnost recent job's executionDuration in JobDispatcher and getExecutionDuration just returns that value? So then regardless when prometheus polls there's always a valid duration value to return for?

Reply:
Prom will gather those values periodically, by default I think it's 15sec but we could increase it if we prefer to improve resolution.
A gauge should gather the current value. If our metric is how long the current job has been running and the current job is null, that means currently the job is not running so it indeed has a running time of 0.
Alerting should be fine as long as duration we alert on is greater than scraping period. E.g. if we want to alert on jobs running >1min this should not be a problem.

Reply:
I think 99% of the time there is no job being run when Prom polls so the gauge will always get 0 which is not informative? and we aren't synchronising `synchronized (jobLock)` inside getExecutionDuration which could create concurrency issue?

So if we use the gauge purely to detect hanged jobs then this is indeed ok tho i thought the gauge can be used as a metric for dashboard too (to see how the duration changes over time, e.g. months as more sites and operators get created). So getting 0 all the time for the gauge isn't useful and probably we need a separate gauge to check for lastSuccessfulRunDuration for us to plot and alert on increase in exec duration?

Reply:
Assigning currentJob to a temp variable shouldn't cause any concurrency issue, especially since the function only uses a getter function. However, I've changed it back to a synchronized lock

Reply:
And execution duration = 0 can just be interpreted as something is not running. If needed, on the same graph visualising the execution durations, we could also query from the logs when a job actually started executing, so we can get points together with the executionDuration line chart

Reply:
as long as the end result is we get an alert if there's no successful job run in last ~25 mins then it's all fine regardless how you implement it. 

so atm we run a job every 20 mins, each time takes around 7-10 seconds so our of an hour (3600 seconds), only 30 out of 3600 seconds the execution duration is not null, so unless Prom polls every second (which I doubt?) otherwise Prom will most likely get 0 most of the time. If it's just for detecting long running/hanged job then it's all good.